metals/src/main/scala/scala/meta/internal/metals/FoldingRangeProvider.scala

```diff
@@ -0,0 +1,82 @@
+package scala.meta.internal.metals
+
+import java.util
+import java.util.Collections
+import org.eclipse.lsp4j.FoldingRange
+import org.eclipse.lsp4j.FoldingRangeCapabilities
+import scala.meta.Template
+import scala.meta.Term
+import scala.meta.Tree
+import scala.meta._
+import scala.meta.inputs.Position
+import scala.meta.internal.metals.FoldingRangeProvider._
+import scala.meta.io.AbsolutePath
+
+final class FoldingRangeProvider(val trees: Trees, foldOnlyLines: Boolean) {
+  def getRangedFor(path: AbsolutePath): util.List[FoldingRange] = {
+    trees
+      .get(path)
+      .map(findFoldingRanges)
+      .getOrElse(util.Collections.emptyList())
+  }
+
+  private def findFoldingRanges(tree: Tree): util.List[FoldingRange] = {
+    val ranges = new FoldingRanges(foldOnlyLines)
+
+    tree traverse {
+      case block: Term.Block => ranges.add(Region, block.pos)
```

Comment:
I am not 100% sure if this is cleaner than the previous explicit Traverser.
here, we need to specify the names like 
```scala
tree traverse {
  case block: Term.Block => foo(block.pos)
  case template: Template => ranges.add(Region, template.pos)
  ...
}
```

but since, we have `tree` in scope - it is easy to make a mistake

in contrast - using explicit Traverser, we had to write:
```scala
def fold(tree: Tree): Unit = tree match {
  case _ : Term.Block => ranges.add(Region, tree.pos)
  case _ : Template => ranges.add(Region, tree.pos)
  case _ =>
}
```

Reply:
You can also do
```scala
x.traverse {
  case tree =>
    tree match {
      case _: Term.Block => ...
```
Either way I don't feel strongly about this, I suspect down the road we will need a custom traverser to stop visiting children so that approach might be more flexible.